demos/twa-post-message/src/main/java/com/google/androidbrowserhelper/demos/twapostmessage/MainActivity.java

```diff
@@ -0,0 +1,122 @@
+          super.onPostMessage(message, extras);
+          if (message.contains("ACK")) {
+            return;
+          }
+          MessageNotificationHandler.showNotificationWithMessage(MainActivity.this, message);
+          Log.d(TAG, "Got message: " + message);
+        }
+
+        @Override
+        public void onRelationshipValidationResult(int relation, @NonNull Uri requestedOrigin,
+            boolean result, @Nullable Bundle extras) {
+          // If this fails:
+          // - Have you called warmup?
+          // - Have you set up Digital Asset Links correctly?
+          // - Double check what browser you're using.
+          Log.d(TAG, "Relationship result: " + result);
+          mValidated = result;
+        }
+
+        // Listens for any navigation happens, we wait until the navigation finishes then we
+        // request post message channel using
+        // CustomTabsSession#requestPostMessageChannel(sourceUri, targetUri, extrasBundle)
+        @Override
+        public void onNavigationEvent(int navigationEvent, @Nullable Bundle extras) {
+          if (navigationEvent != NAVIGATION_FINISHED) {
+            return;
+          }
+
+          if (!mValidated) {
+            Log.d(TAG, "Not starting PostMessage as validation didn't succeed.");
+          }
+
+          // If this fails:
+          // - Have you included PostMessageService in your AndroidManifest.xml ?
+          boolean result = mSession.requestPostMessageChannel(SOURCE_ORIGIN, TARGET_ORIGIN,
+              new Bundle());
+          Log.d(TAG, "Requested Post Message Channel: " + result);
+        }
+
+        @Override
+        public void onMessageChannelReady(@Nullable Bundle extras) {
+          Log.d(TAG, "Message channel ready.");
+
+          int result = mSession.postMessage("First message", null);
+          Log.d(TAG, "postMessage returned: " + result);
+        }
+      };
+
+
+  private void bindCustomTabsService() {
+    String packageName = CustomTabsClient.getPackageName(this, null);
+    Log.d(TAG, "Binding to $packageName.");
+    CustomTabsClient.bindCustomTabsService(this, packageName, new CustomTabsServiceConnection() {
+      @Override
+      public void onCustomTabsServiceConnected(@NonNull ComponentName name,
+          @NonNull CustomTabsClient client) {
+        mClient = client;
+
+        // Note: validateRelationship requires warmup to have been called.
+        client.warmup(0L);
+
+        mSession = mClient.newSession(customTabsCallback);
+        registerBroadcastReceiver();
+      }
+
+      @Override
+      public void onServiceDisconnected(ComponentName componentName) {
+        mClient = null;
+
+      }
+    });
+  }
+
+  private void registerBroadcastReceiver() {
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(PostMessageBroadcastReceiver.POST_MESSAGE_ACTION);
+    registerReceiver(new PostMessageBroadcastReceiver(mSession), intentFilter);
```

Comment:
You need to specify whether the receiver is exported or not, otherwise this will break on Android U+.

https://developer.android.com/about/versions/14/behavior-changes-14#runtime-receivers-exported
https://developer.android.com/guide/components/broadcasts#context-registered-receivers